### ISSUE TITLE
fast coord mapping

### ISSUE THREAD

**Issue by tonghe90** [StanfordVL/MinkowskiEngine#90]
Thanks for sharing the code. I am doing the segmentation problem, but I don' t want to implement on full size. The output of the last layer is, 8s (stride : 8). Given the full-size ground truth, how can I get the indices of the 8s coords in the full size? Do you have any suggestions for fast calculation or is there any API or example that achieves this?

**Comment by chrischoy** [comment]
You can get the coordinates using 

```
coords_8 = ME.utils.sparse_quantize(coords, quantization_size=8)
```

If you want to define labels with `ignore_label` when there are label collision, feed labels into the function as well.  https://stanfordvl.github.io/MinkowskiEngine/utils.html#MinkowskiEngine.utils.sparse_quantize

**Comment by tonghe90** [comment]
Thanks for your reply. But I want to get the matching pairs. For example, I have three conv2s layers and the output stride is 8. Suppose I have the full-size ground truth labels, how can I get the labels for the 8s feature.

**Comment by tonghe90** [comment]
The coords by directly using ME.utils.sparse_quantize(coords, quantization_size=8) do not coincide with the coords after three conv2s layers.

for example, 
coords, feats, labels = ME.utils.sparse_quantize(coords0, features0, labels=instance_labels, ignore_label=0)
sinput = ME.uitls.sparse_collate([coords], [feats], [labels])
soutput = conv1s2(sinput)
soutput = conv2s2(soutput)
soutput = conv3s2(soutput)
the soutput.C is not equal to the ME.utils.sparse_quantize(coords0, features0,labels=instance_labels, ignore_label=0, quantization_size=8)[0]

**Comment by tonghe90** [comment]
How Can I get the ground truth labels for soutput -- 8s

**Comment by chrischoy** [comment]
One way to specify the coordinates and the specified order is to use the convolution forward with coord argument (coords key, sparse tensor).

https://stanfordvl.github.io/MinkowskiEngine/convolution.html#MinkowskiEngine.MinkowskiConvolution.forward

```
criterion = torch.nn.BCELoss(ignore_label=-100)
coords_s8, labels_s8 = ME.utils.sparse_quantize(coords0, features0,labels=instance_labels, ignore_label=-100, quantization_size=8)

sinput = ME.uitls.sparse_collate([coords], [feats], [labels])

sout = conv1s2(sinput)
sout = conv2s4(sout)
sout = conv3s4(sout, coords_s8)  # <-- guarantees that the order matches that of labels_s8
```

**Comment by tonghe90** [comment]
Thanks

**Comment by tonghe90** [comment]
One more thing, if I specified the output coords, is the stride necessary in the init function?

**Comment by chrischoy** [comment]
If you do not specify, the tensor stride will be 1 by default. This does not cause any problem in the above example, but if you wanted to use conv transpose and upsample, this will give you an error.

**Comment by tonghe90** [comment]
I am a little confused. If the output coords are specified, it contains stride information. the kernels are placed in these coords. Why the stride is needed in the conv definition.

**Comment by chrischoy** [comment]
So on the line

```
sout = conv3s2(sout, coords_s8) 
```

coords_s8 is actually an IntTensor. An int tensor itself does not contain any meta data and since you are specifying an output coords without a metadata, it generates a sparse tensor with default stride size which is 1.

However, during the convolution, the tensor stride of the sparse tensor is used correctly to generate the correct output.

To specify the tensor stride, you can follow this example: https://github.com/StanfordVL/MinkowskiEngine/blob/master/examples/convolution.py#L133

**Comment by tonghe90** [comment]
Here is my abstract implementation:
mesh_vertices = np.load('a.npy').astype(np.float32)
mesh_vertices[:, :3] = mesh_vertices[:, :3] - np.min(mesh_vertices[:, :3], axis=0)
instance_labels = np.load('b.npy').astype(np.int32)
coords = np.floor(mesh_vertices[:, :3] / 0.02) ### voxel size
features = mesh_vertices[:, 3:6]

coords_sq_1s, features_sq_1s, labels_sq_1s = ME.utils.sparse_quantize(coords, feats=features, labels=instance_labels, quantization_size=1, ignore_label=0)
coords_sc_1s, feats_sc_1s = ME.utils.sparse_collate([coords_sq_1s], [features_sq_1s])

coords_sq_2s, labels_sq_2s = ME.utils.sparse_quantize(coords, labels=instance_labels, quantization_size=2, ignore_label=0)
coords_sq_2s = coords_sq_2s * 2
coords_sc_2s, _ = ME.utils.sparse_collate([coords_sq_2s], feats=[coords_sq_2s])

coords_sq_4s, labels_sq_4s = ME.utils.sparse_quantize(coords, labels=instance_labels, quantization_size=4, ignore_label=0)
coords_sq_4s = coords_sq_4s * 4
coords_sc_4s, _ = ME.utils.sparse_collate([coords_sq_4s], feats=[coords_sq_4s])

coords_dict = {'1s': torch.IntTensor(coords_sc_1s), '2s': torch.IntTensor(coords_sc_2s), '4s': torch.IntTensor(coords_sc_4s)}

sinput = SparseTensor(feats_sc_1s, coords_sc_1s).to('cuda')

conv1 = conv(stride=1, ...)
conv2 = conv(stride=2, ...)
conv3 = conv(stride=2, ...)

out = conv1(sinput, coords_dict['1s'])
out = conv2(out, coords_dict['2s'])
out = conv3(out, coords_dict['4s'])

CE(out, labels_sq_4s)

Is this correct?

**Comment by chrischoy** [comment]
You don't need to specify the intermediate coordinates tensor stride 1s, 2s. The only reason you want to specify the 4s is to compute the loss. Also, you need to sample points densely from a mesh. Otherwise it would be a very sparse point cloud.
